### PR TITLE
ppo

### PR DESCRIPTION
**Pull request by zhog96**
Start of pop implementation and fixing connected things

### CODE REVIEW

File: agents/ppo/__init__.py [hunk 1]
```diff
@@ -0,0 +1 @@
+from .ppo import PPOAgent, BaseActor, BaseCritic
```
Comment (NonameUntitled) [inline agents/ppo/__init__.py:1]: From my view `ppo/ppo.py` is a bad filename. It gives redundant or ubiquitous explanation. My idea: let's make either a single `ppo.py` file inside `agents` directory or create separated files for actor and critic respectively.

I suggest use the first option because for sure `types.py` is not a part of PPO agent. It should be put in the utils or types section.

File: ml/encoders.py [hunk 2]
```diff
@@ -3,7 +3,7 @@
 
 from typing import List
 
-from . import BaseEncoder, BaseDistance
+from . import BaseEncoder, BaseDistance, TensorWithMask
```
Comment (NonameUntitled) [inline ml/encoders.py:6]: For now imported but unused...

```suggestion
from . import BaseEncoder, BaseDistance
```

File: ml/encoders.py [hunk 3]
```diff
@@ -72,10 +72,10 @@ def forward(self, inputs):
 class SubtractionEncoder(TorchEncoder, config_name='subtraction'):
 
     def __init__(
```
Comment (NonameUntitled) [inline ml/encoders.py:74]: Let's follow PEP8 code style. I will post my PyCharm settings somewhere sometimes...

File: agents/reinforce.py [hunk 4]
```diff
@@ -1,4 +1,5 @@
-from base import TorchAgent
+from ml import BaseEncoder
+from .common import TorchAgent
```
Comment (NonameUntitled) [inline agents/reinforce.py:2]: Same issue as before with usage of relative imports

```suggestion
from agents.common import TorchAgent
```

File: agents/ppo/types.py [hunk 5]
```diff
@@ -0,0 +1,41 @@
...
+
+class Policy(TensorLike, ABC):
+    pass
+
+
+class State(ABC):
+    pass
+
+
+class BaseActor(TorchAgent, config_name="base_actor"):
```
Comment (NonameUntitled) [inline agents/ppo/types.py:32]: Let's use unary parentheses by default:

```suggestion
class BaseActor(TorchAgent, config_name='base_actor'):
```

File: agents/ppo/types.py [hunk 6]
```diff
@@ -0,0 +1,41 @@
...
+    @abstractmethod
+    def act(self, state: State) -> Tuple[State, Policy]:
+        pass
+
+
+class BaseCritic(TorchAgent, config_name="base_critic"):
```
Comment (NonameUntitled) [inline agents/ppo/types.py:38]: Same issue with unary parentheses:

```suggestion
class BaseCritic(TorchAgent, config_name='base_critic'):
```

File: agents/ppo/types.py [hunk 7]
```diff
@@ -0,0 +1,41 @@
...
+    pass
+
+
+class State(ABC):
+    pass
+
+
+class BaseActor(TorchAgent, config_name="base_actor"):
+    @abstractmethod
+    def act(self, state: State) -> Tuple[State, Policy]:
```
Comment (NonameUntitled) [inline agents/ppo/types.py:34]: Let's use here `def forward` since we work only in PyTorch framework, where every encoder should be invoked with `__call__` or `forward` function call.

File: agents/ppo/types.py [hunk 8]
```diff
@@ -0,0 +1,41 @@
...
+        pass
+
+
+class BaseCritic(TorchAgent, config_name="base_critic"):
+    @abstractmethod
+    def value(self, state: State) -> Value:
```
Comment (NonameUntitled) [inline agents/ppo/types.py:40]: Same issue with `def value` vs `def forward`

File: agents/ppo/ppo.py [hunk 9]
```diff
@@ -0,0 +1,52 @@
...
+            ratio_clip=config['ratio_clip']
+        )
+
+    def forward(self, state: State) -> Tuple[State, Policy, Value, Value]:
+        next_state, policy = self._actor.act(state)
+        curr_v_func = self._critic.value(state)
+        q_estimate = self._critic.value(next_state)
+        return next_state, policy, curr_v_func, q_estimate
+
+    def loss(
```
Comment (NonameUntitled) [inline agents/ppo/ppo.py:29]: Let's separate `def loss` on different functions based their usage. From this code it is not true, that all this function is used only for loss computation. Some lines of code should be put into functions like  `def compute_advantage_value` for example

File: agents/ppo/ppo.py [hunk 10]
```diff
@@ -0,0 +1,52 @@
...
+        self._critic = critic
+        self._discount_factor = discount_factor
+        self._ratio_clip = ratio_clip
+
+    @classmethod
+    def create_from_config(cls, config):
+        return cls(
+            actor=BaseActor.create_from_config(config['actor']),
+            critic=BaseCritic.create_from_config(config['critic']),
+            discount_factor=config['discount_factor'],
```
Comment (NonameUntitled) [inline agents/ppo/ppo.py:19]: Let's make params like `discount_factor` and `ratio_clip` parameters with default value and write here something like: `discount_factor=config.get('discount_factor', <default_value>),`

File: agents/ppo/ppo.py [hunk 11]
```diff
@@ -0,0 +1,52 @@
+from agents.ppo.types import *
+from typing import List
+import torch
+
+
+class PPOAgent(TorchAgent, config_name='ppo'):
+    def __init__(self, actor: BaseActor, critic: BaseCritic, discount_factor: float, ratio_clip: float):
```
Comment (NonameUntitled) [inline agents/ppo/ppo.py:7]: There are more than three arguments in this line. Let's re-write it more or less like this:

```suggestion
    def __init__(
        self, actor: BaseActor, 
        critic: BaseCritic, 
        discount_factor: float, 
        ratio_clip: float
    ):
```

File: agents/ppo/ppo.py [hunk 12]
```diff
@@ -0,0 +1,52 @@
...
+        end_v_old_tensor = end_v_old.to_tensor()
+        advantage = start_v_old_tensor
+        for reward in rewards:
+            advantage = self._discount_factor * advantage + reward.to_tensor()
+        advantage -= end_v_old_tensor
+        weighted_prob = advantage * prob_ratio
+        weighted_clipped_prob = torch.clamp(prob_ratio, 1 - self.ratio_clip, 1 + self.ratio_clip) * advantage
+        actor_loss = -torch.min(weighted_prob, weighted_clipped_prob)
+        critic_loss = (advantage + start_v_old_tensor - self._critic.value(state).to_tensor()) ** 2
+        return actor_loss, critic_loss
```
Comment (NonameUntitled) [inline agents/ppo/ppo.py:52]: Let's use same interface for all the losses. For example, I expect from loss function get take something as an input and return single-element float tensor with loss value. Here, for example you can come up with composite/weighted loss

**Review by NonameUntitled — CHANGES_REQUESTED** [review]
Left several comments, let's fix/discuss them and then merge this PR.

**Comment by NonameUntitled** [comment]
Also, it is better to change Name of a PR and Description of a PR to give it more sense

**Comment by NonameUntitled** [comment]
From my point is it better to have only `ppo.py` as a file without package. You can put there specific agent code (`class PPOAgent`), maybe several specific utility functions and etc. 

Why so? Now in this package you have only three files: 
- `ppo.py` (The implementation ppo, you can left it unchanged)
- `types.py` (It is better to make it common for all the agents)
- `__init__.py` It's just because of the existence of a ppo dir

**Comment by NonameUntitled** [comment]
Also, can you please remove `__pycache__` leftovers?

File: agents/ppo/__init__.py (hunk 1 above)
Comment (zhog96) [reply agents/ppo/__init__.py:1]: I agree that `types.py` should be on top level, I think that we may have several implementations of actor, critic and dto classes in one file) may be java code style talks in me and in our case this is ok)

**Review by NonameUntitled — APPROVED** [review]

File: agents/ppo.py [hunk 13]
```diff
@@ -0,0 +1,60 @@
...
+        end_v_old_tensor = end_v_old.to_tensor()
+        advantage = start_v_old_tensor
+        for reward in rewards:
+            advantage = self._discount_factor * advantage + reward.to_tensor()
+        advantage -= end_v_old_tensor
+        weighted_prob = advantage * prob_ratio
+        weighted_clipped_prob = torch.clamp(prob_ratio, 1 - self.ratio_clip, 1 + self.ratio_clip) * advantage
+        actor_loss = -torch.min(weighted_prob, weighted_clipped_prob)
+        critic_loss = (advantage + start_v_old_tensor - self._critic.forward(state).to_tensor()) ** 2
+        return actor_loss, critic_loss
```
Comment (NonameUntitled) [inline agents/ppo.py:60]: The same issue about the same interface for all losses

File: agents/ppo.py [hunk 14]
```diff
@@ -0,0 +1,60 @@
...
+            ratio_clip=config.get('ratio_clip', 0.2)
+        )
+
+    def forward(self, state: State) -> Tuple[State, Policy, Value, Value]:
+        next_state, policy = self._actor.forward(state)
+        curr_v_func = self._critic.forward(state)
+        q_estimate = self._critic.forward(next_state)
+        return next_state, policy, curr_v_func, q_estimate
+
+    def loss(
```
Comment (NonameUntitled) [inline agents/ppo.py:37]: Let's separate `def loss` on different functions based their usage. From this code it is not true, that all this function is used only for loss computation. Some lines of code should be put into functions like  `def compute_advantage_value` for example

File: agents/ppo.py [hunk 15]
```diff
@@ -0,0 +1,60 @@
...
+            rewards: List[Reward],
+            end_v_old: Value,
+    ) -> None:
+        _, policy = self._actor.forward(state)
+        policy_tensor = policy.to_tensor()
+        policy_old_tensor = policy_old.to_tensor()
+        prob_ratio = policy_tensor / policy_old_tensor
+        start_v_old_tensor = start_v_old.to_tensor()
+        end_v_old_tensor = end_v_old.to_tensor()
+        advantage = start_v_old_tensor
```
Comment (NonameUntitled) [inline agents/ppo.py:52]: For instance here you can implement `def compute_advantage_score` inside `class PPOAgent`

**Review by NonameUntitled — APPROVED** [review]
